staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1154 @@
+	if err != nil {
+		return err
+	}
+	glog.V(2).Infoln("Local NetworkAttachments cache synced")
+
+	err = ca.handlePreExistingIfcs()
+	if err != nil {
+		return err
+	}
+	glog.V(2).Infoln("Pre-existing interfaces synced")
+
+	for i := 0; i < ca.workers; i++ {
+		go k8swait.Until(ca.processQueue, time.Second, stopCh)
+	}
+	glog.V(2).Infof("Launched %d workers", ca.workers)
+
+	<-stopCh
+	return nil
+}
+
+func (ca *ConnectionAgent) initLocalAttsInformerAndLister() {
+	localAttWithAnIPSelector := ca.localAttWithAnIPSelector()
+	glog.V(6).Info("Created Local NetworkAttachments fields selector: " +
+		localAttWithAnIPSelector)
+
+	ca.localAttsInformer, ca.localAttsLister = v1a1AttsCustomInformerAndLister(ca.kcs,
+		resyncPeriod,
+		fromFieldsSelectorToTweakListOptionsFunc(localAttWithAnIPSelector))
+
+	ca.localAttsInformer.AddIndexers(map[string]k8scache.IndexFunc{attMACIndexName: attachmentMACAddr})
+
+	ca.localAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    ca.onLocalAttAdded,
+		UpdateFunc: ca.onLocalAttUpdated,
+		DeleteFunc: ca.onLocalAttRemoved,
+	})
+}
+
+func (ca *ConnectionAgent) onLocalAttAdded(obj interface{}) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of addition of %#+v", att)
+	ca.queue.Add(kosctlrutils.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) onLocalAttUpdated(oldObj, newObj interface{}) {
+	oldAtt := oldObj.(*netv1a1.NetworkAttachment)
+	newAtt := newObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of update from %#+v to %#+v",
+		oldAtt,
+		newAtt)
+	ca.queue.Add(kosctlrutils.AttNSN(newAtt))
+}
+
+func (ca *ConnectionAgent) onLocalAttRemoved(obj interface{}) {
+	peeledObj := kosctlrutils.Peel(obj)
+	att := peeledObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of removal of %#+v", att)
+	ca.queue.Add(kosctlrutils.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) waitForLocalAttsCacheSync(stopCh <-chan struct{}) (err error) {
+	if !k8scache.WaitForCacheSync(stopCh, ca.localAttsInformer.HasSynced) {
+		err = fmt.Errorf("Caches failed to sync")
+	}
+	return
+}
+
+func (ca *ConnectionAgent) handlePreExistingIfcs() (err error) {
+	err = ca.handlePreExistingLocalIfcs()
+	if err != nil {
+		return
+	}
+
+	err = ca.handlePreExistingRemoteIfcs()
+	return
+}
+
+func (ca *ConnectionAgent) handlePreExistingLocalIfcs() error {
+	allPreExistingLocalIfcs, err := ca.netFabric.ListLocalIfcs()
+	if err != nil {
+		return fmt.Errorf("Failed initial local network interfaces list: %s", err.Error())
+	}
+
+	for _, aPreExistingLocalIfc := range allPreExistingLocalIfcs {
+		ifcMAC := aPreExistingLocalIfc.GuestMAC.String()
+		ifcOwnerAtts, err := ca.localAttsInformer.GetIndexer().ByIndex(attMACIndexName, ifcMAC)
+		if err != nil {
+			return fmt.Errorf("Indexing local network interface with MAC %s failed: %s",
+				ifcMAC,
+				err.Error())
+		}
+
+		if len(ifcOwnerAtts) == 1 {
```

Comment:
Can `len(ifcOwnerAtts)` be greater than 1?  If yes then this case should be handled.  If not, then perhaps there should be a comment here describing why that is impossible.

I think `len(ifcOwnerAtts)` is certainly not greater than 1.  Following is a proof by contradiction.  Suppose it is two.  That means the index of the local informer said two local NetworkAttachment objects have the same MAC address.  MAC address is a function of VNI and IP address.  The IPAM controller does not allow such a collision to exist at any ResourceVersion of NetworkAttachment.  The index could have a collision only if it is notified of the second binding of the VNI&MAC before being notified of the first unbinding.  But an informer will not make such out-of-order notifications.

Reply:
1. `len(ifcOwnerAtts)` can be 0, in such case we have to delete the interface, hence a check like the one in the code is needed.
2. You're forgetting about the usual problem: updates to the VNI field. Consider this scenario: att1 and att2, with VNI1 and VNI2 respectively, have the same IP. att2 is updated by changing its VNI from 2 to 1. The IPAM controller will react by releasing the old lock and allocating a new one, but it takes time. In the meantime, the Informer's cache could store two attachments with the same MAC, hence the check on len. What the code currently does in this scenario is simply deleting the interface, and hope that by the time the reference of att2 is processed the Informer's cache already stores the version with the new IP, hence the MACs of the interfaces which get created differ. This is not guaranteed to succeed however. Do I have a plan to fix this? Yes. I see two options: (A) add a field to the status of an attachment with the name of its IPLock. This field is set by the IPAM when it sets the UID of the lock (which is already doing). When a connection agent worker retrieves an attachment, it stops processing it if the VNI in the spec differs from the VNI in the name of its IPlock. When the attachment is updated to reflect that it has a new (legitimate) address, its reference will be enqueued again. This is the solution I favor since code-wise it requires tiny additions both to the IPAM controller and the connection agent. (B) If we want to keep the types unchanged, we can have the ConnectionAgent maintain a cache which associates VNIs to IP addresses to attachments seen with such IP addresses. A worker always adds the tuple (VNI, Namespaced Name, IP) to the cache when processing an existing attachment, and removes it when deleting it (and if there's only another attachment which has been seen with that IP it enqueues its reference). If it sees that the IP address of the attachment is processing is already of a different Attachment (in the cache) it stops processing (when things change and this ambiguity has been resolved a reference will be enqueued). I don't like this solution because it adds considerable complexity ConnectionAgent-side, and it's less correct than the one with the additional field in the status of the NetworkAttachment: it only guarantees that no two interfaces with the same MAC are created. But in the case of an update to the VNI field which leads to an attachment which temporarily holds an IP with a lock from another VNI, it does not guarantee that the interface which gets created is the one associated with the rightful owner of the IP and the interface. For a short transient, the attachment with the lock from the old VNI might have the interface, and keep the rightful owner of such interface from creating it (as soon as the "illegitimate" owner gets the new IP this is resolved).

Reply:
Yes, it is clear that the index might find zero objects.
You are right, a change in VNI could make the index find two objects.  I agree with your preferred approach to solving the problem, but suggest this twist: make the Status store the VNI, rather than the lock name, associated with the IP address.  That makes NetworkAttachmentStatus independent of the implementation details of the IPAM controller.

Reply:
I suggest indexing on <Status.VNI, Status.IPv4> rather than on Status.IfcName, because the former is set earlier than the latter.  Consider this scenario.  NetworkAttachment object is created, then gets an IP address assigned, then its local connection agent propagates it into the network fabric, then attempts to update the NetworkAttachmentStatus but fails, then that connection agent crashes and restarts.  In the following startup, the matching NetworkAttachment will be found if sought by <Status.VNI, Status.IPv4> but not if sought by Status.IfcName.

Reply:
Yes, setting the VNI is better, you're right. The current Index is based on the MAC address, which is a function of <Status.VNI, Status.IPv4>. IfcName is not used.